### PR TITLE
Fix bug with incorrectly defactorized dependencies #768

### PR DESCRIPTION
**Pull request by bartsanchez**
New version with improved changes, fixing errors raised
with the previous version:

https://github.com/tox-dev/tox/issues/899
https://github.com/tox-dev/tox/issues/906

## Thanks for contributing a pull request!

If you are contributing for the first time or provide a trivial fix don't worry too
much about the checklist - we will help you get started.

## Contribution checklist:

(also see [CONTRIBUTING.rst](https://github.com/tox-dev/tox/tree/master/CONTRIBUTING.rst) for details)

- [ ] wrote descriptive pull request text
- [ ] added/updated test(s)
- [ ] updated/extended the documentation
- [ ] added relevant [issue keyword](https://help.github.com/articles/closing-issues-using-keywords/)
      in message body
- [ ] added news fragment in [changelog folder](https://github.com/tox-dev/tox/tree/master/docs/changelog)
  * fragment name: `<issue number>.<type>.rst` for example (588.bugfix.rst)
  * `<type>` is must be one of `bugfix`, `feature`, `deprecation`,`breaking`, `doc`, `misc`
  * if pr has no issue: consider creating one first or change it to the pr number after creating the pr
  * "sign" fragment with "by @<your username>"
  * please use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files - by :user:`superuser`."
  * also see [examples](https://github.com/tox-dev/tox/tree/master/changelog/examples.rst)
- [ ] added yourself to `CONTRIBUTORS` (preserving alphabetical order)

### CODE REVIEW

File: src/tox/config.py [hunk 1]
```diff
@@ -1337,10 +1337,16 @@ def getstring(self, name, default=None, replace=True, crossonly=False):
             x = default
         else:
             x = self._apply_factors(x)
+            x = self._replace_if_needed(x, name, replace, crossonly)
+            x = self._apply_factors(x)
 
+        x = self._replace_if_needed(x, name, replace, crossonly)
+        # print "getstring", self.section_name, name, "returned", repr(x)
+        return x
```
[the remark this answers is not in the document]
Comment (bartsanchez) [reply src/tox/config.py]: The comment is not mine, I kept it as it was already in the code.

I would prefer to not having commented code. If you do prefer I can simply remove it.

File: tests/unit/test_config.py [hunk 2]
```diff
@@ -1477,6 +1475,226 @@ def test_regression_test_issue_706(self, newconfig, envlist):
...
+        conf = newconfig([], inisource).envconfigs["py27-cover"]
+        packages = [dep.name for dep in conf.deps]
+        assert packages == ["foo"]
+
+        conf = newconfig([], inisource).envconfigs["py37-cover"]
+        packages = [dep.name for dep in conf.deps]
+        assert packages == ["bar"]
+
+    # https://github.com/tox-dev/tox/issues/899
+    def test_regression_test__issue_899(self, newconfig):
```
[the remark this answers is not in the document]
Comment (bartsanchez) [reply tests/unit/test_config.py]: OK, I'll change it.

File: tests/unit/test_config.py [hunk 3]
```diff
@@ -1477,6 +1475,226 @@ def test_regression_test_issue_706(self, newconfig, envlist):
...
+                style
+                sdist
+                bdist_wheel
+                {py27,py34,py35,py36,pypy,pypy3}-cover
+                {py27,py34,py35,py36,pypy,pypy3}-nocov
+
+            [testenv]
+            skip_install = True
+            sitepackages = True
+            passenv =
```
[the remark this answers is not in the document]
Comment (bartsanchez) [reply tests/unit/test_config.py]: OK, I'll change it.

File: tests/unit/test_config.py [hunk 4]
```diff
@@ -1477,6 +1475,226 @@ def test_regression_test_issue_706(self, newconfig, envlist):
...
+        assert packages == ["rdflib", "numpy==1.12.1"]
+
+        conf = newconfig([], inisource).envconfigs["pypy3-cover"]
+        packages = [dep.name for dep in conf.deps]
+        assert packages == [
+            "coverage", "codecov", "numpy==1.12.1",
+        ]
+
+    # https://github.com/tox-dev/tox/issues/906
+    def test_regression_test__issue_906(self, newconfig):
```
[the remark this answers is not in the document]
Comment (bartsanchez) [reply tests/unit/test_config.py]: OK, I'll change it.

File: src/tox/config.py [hunk 5]
```diff
@@ -1337,10 +1337,16 @@ def getstring(self, name, default=None, replace=True, crossonly=False):
             x = default
         else:
             x = self._apply_factors(x)
+            x = self._replace_if_needed(x, name, replace, crossonly)
+            x = self._apply_factors(x)
 
+        x = self._replace_if_needed(x, name, replace, crossonly)
```
[the remark this answers is not in the document]
Comment (bartsanchez) [reply src/tox/config.py]: OK, I'll try to summarize the issue as much as I can.

File: src/tox/config.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (gaborbernat) [reply src/tox/config.py]: thanks

**Pull request closed without merging**